sdk/storage/Azure.Storage.Common/src/Shared/StructuredMessageEncodingStream.cs

```diff
@@ -0,0 +1,525 @@
+            ArrayPool<byte>.Shared,
+            out Memory<byte> headerBytes,
+            CurrentInnerSegment,
+            Math.Min(_segmentContentLength, _innerStream.Length - _innerStream.Position));
+        headerBytes.Slice(_currentRegionPosition, read).Span.CopyTo(buffer);
+        _currentRegionPosition += read;
+
+        if (_currentRegionPosition == _segmentHeaderLength)
+        {
+            _currentRegion = SMRegion.SegmentContent;
+            _currentRegionPosition = 0;
+        }
+
+        return read;
+    }
+
+    private int ReadFromSegmentFooter(Span<byte> buffer)
+    {
+        int read = Math.Min(buffer.Length, _segmentFooterLength - _currentRegionPosition);
+        if (read < 0)
+        {
+            return 0;
+        }
+
+        using IDisposable _ = StructuredMessage.V1_0.GetSegmentFooterBytes(
+            ArrayPool<byte>.Shared,
+            out Memory<byte> headerBytes,
+            crc64: UseCrcSegment
+                ? new Span<byte>(
+                    _runningCrcCheckpoints,
+                    (CurrentEncodingSegment-1) * _runningCrc.HashLengthInBytes,
+                    _runningCrc.HashLengthInBytes)
+                : default);
+        headerBytes.Slice(_currentRegionPosition, read).Span.CopyTo(buffer);
+        _currentRegionPosition += read;
+
+        if (_currentRegionPosition == _segmentFooterLength)
+        {
+            _currentRegion = _innerStream.Position == _innerStream.Length
+                ? SMRegion.StreamFooter : SMRegion.SegmentHeader;
+            _currentRegionPosition = 0;
+        }
+
+        return read;
+    }
+    #endregion
+
+    #region ReadUnderlyingStream
+    private int MaxInnerStreamRead => _segmentContentLength - _currentRegionPosition;
+
+    private void CleanupContentSegment()
+    {
+        if (_currentRegionPosition == _segmentContentLength || _innerStream.Position >= _innerStream.Length)
+        {
+            _currentRegion = SMRegion.SegmentFooter;
+            _currentRegionPosition = 0;
+            if (UseCrcSegment && CurrentEncodingSegment - 1 == _latestSegmentCrcd)
+            {
+                _runningCrc.GetCurrentHash(new Span<byte>(
+                    _runningCrcCheckpoints,
+                    _latestSegmentCrcd * _runningCrc.HashLengthInBytes,
+                    _runningCrc.HashLengthInBytes));
+                _latestSegmentCrcd++;
+            }
+        }
+    }
+
+    private async ValueTask<int> ReadFromInnerStreamInternal(
+        byte[] buffer, int offset, int count, bool async, CancellationToken cancellationToken)
+    {
+        int read = async
+            ? await _innerStream.ReadAsync(buffer, offset, Math.Min(count, MaxInnerStreamRead)).ConfigureAwait(false)
+            : _innerStream.Read(buffer, offset, Math.Min(count, MaxInnerStreamRead));
+        _currentRegionPosition += read;
+        CleanupContentSegment();
+        return read;
+    }
+
+#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP3_0_OR_GREATER
```

Comment:
Would we be using something else for .NET Standard 2.0, or can we give it coverage due to limitations?